File: apps/android/app/src/main/java/chat/rogi/rogichat/channelport/core/domain/util/SongPriceCalculator.kt
package chat.rogi.rogichat.channelport.core.domain.util

import chat.rogi.rogichat.channelport.core.model.song.Category
import chat.rogi.rogichat.channelport.core.model.song.CurrencyConfig
import chat.rogi.rogichat.channelport.core.model.song.PricingSettings

enum class PriceSource {
    SONG,
    CATEGORY,
    DIFFICULTY,
    DEFAULT,
    FREE,
}

data class SongPriceInput(
    val price: Int? = null,
    val currencyPrices: Map<String, Int?>? = null,
    val difficulty: Int? = null,
    val categories: List<Category> = emptyList(),
)

data class CalculatedSongPrice(
    val price: Int?,
    val source: PriceSource,
    val currencyKey: String? = null,
)

data class SongPriceItem(
    val currencyKey: String,
    val price: Int?,
    val unit: String,
    val source: PriceSource,
)

object SongPriceCalculator {

    fun calculate(
        song: SongPriceInput,
        pricingSettings: PricingSettings?,
        currencyKey: String? = null,
    ): CalculatedSongPrice {
        // If pricing is disabled or settings are null, return FREE
        if (pricingSettings == null || !pricingSettings.pricingEnabled) {
            return CalculatedSongPrice(price = null, source = PriceSource.FREE, currencyKey = currencyKey)
        }

        val resolvedCurrencyKey = resolveCurrencyKey(song, pricingSettings, currencyKey)

        // Priority 1: Song's own price
        val songPrice = getPriceByCurrencyOrLegacy(
            song.currencyPrices,
            resolvedCurrencyKey,
            song.price
        )
        if (songPrice != null) {
            return CalculatedSongPrice(price = songPrice, source = PriceSource.SONG, currencyKey = resolvedCurrencyKey)
        }

        // Priority 2 & 3: Difficulty and Category (take the higher one)
        val difficultyPrice = getDifficultyPrice(
            song.difficulty,
            resolvedCurrencyKey,
            pricingSettings.difficultyPrices,
            pricingSettings.difficultyPricesByCurrency
        )
        val categoryPrice = getMaxCategoryPrice(song.categories, resolvedCurrencyKey)

        if (difficultyPrice != null || categoryPrice != null) {
            if (difficultyPrice != null && categoryPrice != null) {
                return if (difficultyPrice >= categoryPrice) {
                    CalculatedSongPrice(price = difficultyPrice, source = PriceSource.DIFFICULTY, currencyKey = resolvedCurrencyKey)
                } else {
                    CalculatedSongPrice(price = categoryPrice, source = PriceSource.CATEGORY, currencyKey = resolvedCurrencyKey)
                }
            }
            if (difficultyPrice != null) {
                return CalculatedSongPrice(price = difficultyPrice, source = PriceSource.DIFFICULTY, currencyKey = resolvedCurrencyKey)
            }
            return CalculatedSongPrice(price = categoryPrice, source = PriceSource.CATEGORY, currencyKey = resolvedCurrencyKey)
        }

        // Priority 4: Default price
        val defaultPrice = getPriceByCurrencyOrLegacy(
            pricingSettings.defaultPrices,
            resolvedCurrencyKey,
            pricingSettings.defaultPrice
        )
        if (defaultPrice != null) {
            return CalculatedSongPrice(price = defaultPrice, source = PriceSource.DEFAULT, currencyKey = resolvedCurrencyKey)
        }

        // No price found
        return CalculatedSongPrice(price = null, source = PriceSource.FREE, currencyKey = resolvedCurrencyKey)
    }

    fun getPriceItems(
        song: SongPriceInput,
        pricingSettings: PricingSettings?,
    ): List<SongPriceItem> {
        if (pricingSettings == null || !pricingSettings.pricingEnabled) {
            return emptyList()
        }

        val configs = normalizeCurrencyConfigs(pricingSettings.currencyConfigs)

        // Multi-currency mode
        if (configs.isNotEmpty()) {
            return configs.map { config ->
                val result = calculate(song, pricingSettings, currencyKey = config.key)
                SongPriceItem(
                    currencyKey = config.key,
                    price = result.price,
                    unit = config.unit,
                    source = result.source,
                )
            }
        }

        // Legacy single-currency mode
        val result = calculate(song, pricingSettings)
        if (result.price == null) {
            return emptyList()
        }
        return listOf(
            SongPriceItem(
                currencyKey = "",
                price = result.price,
                unit = pricingSettings.currencyUnit,
                source = result.source,
            )
        )
    }

    // -- Private helpers --

    private fun normalizeCurrencyConfigs(configs: List<CurrencyConfig>?): List<CurrencyConfig> {
        return (configs ?: emptyList()).filter { config ->
            config.key.trim().isNotEmpty() && config.unit.trim().isNotEmpty()
        }
    }

    private fun resolveCurrencyKey(
        song: SongPriceInput,
        pricingSettings: PricingSettings,
        preferredCurrencyKey: String?,
    ): String? {
        if (preferredCurrencyKey != null) {
            return preferredCurrencyKey
        }

        val configs = normalizeCurrencyConfigs(pricingSettings.currencyConfigs)
        if (configs.isNotEmpty()) {
            return configs[0].key
        }

        // Fallback: look for first key in any currency price map
        val priceMaps: List<Map<String, *>?> = buildList {
            add(song.currencyPrices)
            song.categories.forEach { add(it.currencyPrices) }
            add(pricingSettings.defaultPrices)
        }
        for (map in priceMaps) {
            if (map != null) {
                val firstKey = map.keys.firstOrNull()
                if (firstKey != null) {
                    return firstKey
                }
            }
        }

        return null
    }

    private fun getPriceByCurrencyOrLegacy(
        currencyPrices: Map<String, Int?>?,
        currencyKey: String?,
        legacyPrice: Int?,
    ): Int? {
        if (currencyKey != null && currencyPrices != null && currencyPrices.containsKey(currencyKey)) {
            return currencyPrices[currencyKey]
        }
        return legacyPrice
    }

    private fun getDifficultyPrice(
        difficulty: Int?,
        currencyKey: String?,
        difficultyPrices: Map<String, Int?>?,
        difficultyPricesByCurrency: Map<String, Map<String, Int?>?>?,
    ): Int? {
        if (difficulty == null) return null
        val key = difficulty.toString()

        if (currencyKey != null) {
            val byCurrency = difficultyPricesByCurrency?.get(currencyKey)
            if (byCurrency != null && byCurrency.containsKey(key)) {
                return byCurrency[key]
            }
        }

        return difficultyPrices?.get(key)
    }

    private fun getMaxCategoryPrice(
        categories: List<Category>,
        currencyKey: String?,
    ): Int? {
        if (categories.isEmpty()) return null

        val prices = categories.mapNotNull { category ->
            getPriceByCurrencyOrLegacy(category.currencyPrices, currencyKey, category.price)
        }
        if (prices.isEmpty()) return null
        return prices.max()
    }
}

File: apps/ios/Sources/Features/Channel/SongPriceCalculator.swift
import Foundation

enum PriceSource: String, Equatable {
    case song = "SONG"
    case category = "CATEGORY"
    case difficulty = "DIFFICULTY"
    case default_ = "DEFAULT"
    case free = "FREE"
}

struct SongPriceInput {
    let price: Int?
    let currencyPrices: [String: Int?]?
    let difficulty: Int?
    let categories: [Category]
}

struct CalculatedSongPrice {
    let price: Int?
    let source: PriceSource
    let currencyKey: String?
}

struct SongRequestPriceItem {
    let currencyKey: String?
    let unit: String
    let price: Int?
    let source: PriceSource
}

enum SongPriceCalculator {

    static func calculate(
        song: SongPriceInput,
        settings: PricingSettings?,
        currencyKey: String? = nil
    ) -> CalculatedSongPrice {
        guard let settings = settings, settings.pricingEnabled else {
            return CalculatedSongPrice(price: nil, source: .free, currencyKey: currencyKey)
        }

        let resolvedKey = currencyKey ?? resolveCurrencyKey(song: song, settings: settings)

        // Priority 1: Song price
        let songPrice = getPriceByCurrencyOrLegacy(
            currencyPrices: song.currencyPrices,
            currencyKey: resolvedKey,
            legacyPrice: song.price
        )
        if songPrice != nil {
            return CalculatedSongPrice(price: songPrice, source: .song, currencyKey: resolvedKey)
        }

        // Priority 2 & 3: Difficulty and Category (higher wins)
        let difficultyPrice = getDifficultyPrice(
            difficulty: song.difficulty,
            currencyKey: resolvedKey,
            difficultyPrices: settings.difficultyPrices,
            difficultyPricesByCurrency: settings.difficultyPricesByCurrency
        )
        let categoryPrice = getMaxCategoryPrice(
            categories: song.categories,
            currencyKey: resolvedKey
        )

        if difficultyPrice != nil || categoryPrice != nil {
            if let dp = difficultyPrice, let cp = categoryPrice {
                if dp >= cp {
                    return CalculatedSongPrice(price: dp, source: .difficulty, currencyKey: resolvedKey)
                }
                return CalculatedSongPrice(price: cp, source: .category, currencyKey: resolvedKey)
            }
            if let dp = difficultyPrice {
                return CalculatedSongPrice(price: dp, source: .difficulty, currencyKey: resolvedKey)
            }
            return CalculatedSongPrice(price: categoryPrice, source: .category, currencyKey: resolvedKey)
        }

        // Priority 4: Default price
        let defaultPrice = getPriceByCurrencyOrLegacy(
            currencyPrices: settings.defaultPrices,
            currencyKey: resolvedKey,
            legacyPrice: settings.defaultPrice
        )
        if defaultPrice != nil {
            return CalculatedSongPrice(price: defaultPrice, source: .default_, currencyKey: resolvedKey)
        }

        return CalculatedSongPrice(price: nil, source: .free, currencyKey: resolvedKey)
    }

    static func getPriceItems(
        song: SongPriceInput,
        settings: PricingSettings?
    ) -> [SongRequestPriceItem] {
        guard let settings = settings, settings.pricingEnabled else {
            return []
        }

        let configs = settings.currencyConfigs.filter { !$0.key.isEmpty && !$0.unit.isEmpty }

        if !configs.isEmpty {
            let items = configs.map { config in
                let result = calculate(song: song, settings: settings, currencyKey: config.key)
                return SongRequestPriceItem(
                    currencyKey: config.key,
                    unit: config.unit,
                    price: result.price,
                    source: result.source
                )
            }
            let priced = items.filter { $0.price != nil }
            if !priced.isEmpty { return priced }
            return items.isEmpty ? [] : [items[0]]
        }

        let result = calculate(song: song, settings: settings)
        return [SongRequestPriceItem(
            currencyKey: result.currencyKey,
            unit: resolveUnit(currencyKey: result.currencyKey, settings: settings),
            price: result.price,
            source: result.source
        )]
    }

    // MARK: - Private Helpers

    private static func getPriceByCurrencyOrLegacy(
        currencyPrices: [String: Int?]?,
        currencyKey: String?,
        legacyPrice: Int?
    ) -> Int? {
        if let key = currencyKey,
           let prices = currencyPrices,
           prices.keys.contains(key) {
            return prices[key] ?? nil
        }
        return legacyPrice
    }

    private static func getDifficultyPrice(
        difficulty: Int?,
        currencyKey: String?,
        difficultyPrices: [String: Int?]?,
        difficultyPricesByCurrency: [String: [String: Int?]?]?
    ) -> Int? {
        guard let difficulty = difficulty else { return nil }
        let key = String(difficulty)

        if let currencyKey = currencyKey,
           let byCurrency = difficultyPricesByCurrency?[currencyKey],
           let map = byCurrency,
           map.keys.contains(key) {
            return map[key] ?? nil
        }

        if let prices = difficultyPrices, prices.keys.contains(key) {
            return prices[key] ?? nil
        }

        return nil
    }

    private static func getMaxCategoryPrice(
        categories: [Category],
        currencyKey: String?
    ) -> Int? {
        let prices = categories.compactMap { category -> Int? in
            getPriceByCurrencyOrLegacy(
                currencyPrices: category.currencyPrices,
                currencyKey: currencyKey,
                legacyPrice: category.price
            )
        }
        return prices.isEmpty ? nil : prices.max()
    }

    private static func resolveCurrencyKey(
        song: SongPriceInput,
        settings: PricingSettings
    ) -> String? {
        let configs = settings.currencyConfigs.filter { !$0.key.isEmpty && !$0.unit.isEmpty }
        if let first = configs.first { return first.key }

        // Check song currencyPrices
        if let key = song.currencyPrices?.keys.first { return key }
        // Check category currencyPrices
        for cat in song.categories {
            if let key = cat.currencyPrices?.keys.first { return key }
        }
        // Check defaultPrices
        if let key = settings.defaultPrices?.keys.first { return key }

        return nil
    }

    private static func resolveUnit(currencyKey: String?, settings: PricingSettings) -> String {
        if let key = currencyKey,
           let config = settings.currencyConfigs.first(where: { $0.key == key }) {
            return config.unit
        }
        if !settings.currencyUnit.isEmpty { return settings.currencyUnit }
        return ""
    }
}
